### PR TITLE
Gap 2587/redirect to404when not allowed on a page

### PR DESCRIPTION
**Pull request by a-lor-cab**
[GAP-2587](https://technologyprogramme.atlassian.net/browse/GAP-2587)

## Description
if a user was logged in, and tried to access any other subdomain not within his role capabilities, the user service was redirecting the user to its dashboard.
we now implemented logic to redirect the user to a 404/page based on the user role

Summary of the changes and the related issue. List any dependencies that are required for this change:
https://github.com/cabinetoffice/gap-user-service/pull/202

## Type of change

Please check the relevant options.

- [x] Bug fix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] This change requires a documentation update.

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes:

- [x] Unit Test

- [ ] Integration Test (if applicable)

- [x] End to End Test (if applicable)

## Screenshots (if appropriate):

Please attach screenshots of the change if it is a UI change:

# Checklist:

- [ ] If I have listed dependencies above, I have ensured that they are present in the target branch.
- [ ] I have performed a self-review of my code.
- [ ] I have commented my code in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation where applicable.
- [ ] I have ran cypress tests and they all pass locally.

[GAP-2587]: https://technologyprogramme.atlassian.net/browse/GAP-2587?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/admin/src/middleware.page.ts [hunk 1]
```diff
@@ -11,6 +11,16 @@ const authenticateRequest = async (req: NextRequest, res: NextResponse) => {
...
   const userJwtCookie = req.cookies.get(process.env.JWT_COOKIE_NAME);
 
+  await csrfMiddleware(req, res);
+  //means that the user is logged in but not as an admin/superAdmin(otherwise they would have had the session_id cookie set)
+  if (!authCookie && userJwtCookie) {
+    const url = getLoginUrl({ redirectTo404: true });
+    logger.info(
+      `User is not an admin - redirecting it to appropriate app 404 page`
+    );
+    return NextResponse.redirect(url);
```
Comment (ConorFayleTCO) [inline packages/admin/src/middleware.page.ts:21]:
```suggestion
    return NextResponse.redirect(url, { status: 302 });
```
probably best to follow this pattern as with the others

File: packages/admin/src/middleware.page.ts [hunk 2]
```diff
@@ -11,6 +11,16 @@ const authenticateRequest = async (req: NextRequest, res: NextResponse) => {
   const authCookie = req.cookies.get('session_id');
   const userJwtCookie = req.cookies.get(process.env.JWT_COOKIE_NAME);
 
+  await csrfMiddleware(req, res);
+  //means that the user is logged in but not as an admin/superAdmin(otherwise they would have had the session_id cookie set)
+  if (!authCookie && userJwtCookie) {
```
Comment (ConorFayleTCO) [inline packages/admin/src/middleware.page.ts:16]:
```suggestion
  const isLoggedIn = !!userJwtCookie;
  const isLoggedInAsAdmin = !!authCookie
  if (isLoggedIn && !isLoggedInAsAdmin) {
```

File: packages/admin/src/middleware.page.ts [hunk 3]
```diff
@@ -11,6 +11,16 @@ const authenticateRequest = async (req: NextRequest, res: NextResponse) => {
...
+    const url = getLoginUrl({ redirectTo404: true });
+    logger.info(
+      `User is not an admin - redirecting it to appropriate app 404 page`
+    );
+    return NextResponse.redirect(url);
+  }
+  //means user is not logged in
   if (!authCookie || !userJwtCookie) {
```
Comment (ConorFayleTCO) [inline packages/admin/src/middleware.page.ts:24]:
```suggestion
  if (!isLoggedIn || !isLoggedInAsAdmin) {
```

Comment (ConorFayleTCO) [inline packages/admin/src/middleware.page.ts:24]:
```suggestion
  } else if (!isLoggedIn || !isLoggedInAsAdmin) {
```

**Review by tco-olivergodby — APPROVED** [review]
Looks good to me - merge when ready

File: packages/admin/src/middleware.page.ts [hunk 4]
```diff
@@ -11,7 +11,17 @@ const authenticateRequest = async (req: NextRequest, res: NextResponse) => {
   const authCookie = req.cookies.get('session_id');
   const userJwtCookie = req.cookies.get(process.env.JWT_COOKIE_NAME);
 
-  if (!authCookie || !userJwtCookie) {
+  await csrfMiddleware(req, res);
```
Comment (jgunnCO) [inline packages/admin/src/middleware.page.ts:14]: hmm git has been doing some strange merges today - this line needs to be removed or we'll end up calling the middleware twice, which does not work as I discovered yesterday

**Review by jgunnCO — APPROVED** [review]

**Pull request merged**